utils/misc.mli

```diff
@@ -331,3 +331,137 @@ module type HookSig = sig
+      to read to obtain the byte sequence that should be a magic
+      number. Typical user code will look like:
+      {|
+        let ic = open_in_bin path in
+        let magic =
+          try really_input_string ic Magic_number.magic_length
+          with End_of_line -> ... in
+        match Magic_number.parse magic with
+        | Error parse_error -> ...
+        | Ok info -> ...
+      |}
+
+      A given compiler version expects one specific version for each
+      kind of object file, and will fail if given an unsupported
+      version. Because versions grow monotonically, you can compare
+      the parsed version with the expected "current version" for
+      a kind, to tell whether the wrong-magic object file comes from
+      the past or from the future.
+
+      An example of code block that expects the "currently supported version"
+      of a given kind of magic numbers, here [Cmxa], is as follows:
+      {|
+        let ic = open_in_bin path in
+        begin
+          try Magic_number.(expect_current Cmxa (get_info ic)) with
+          | Parse_error error -> ...
+          | Unexpected error -> ...
+        end;
+        ...
+      |}
+
+      Parse errors distinguish inputs that are [Not_a_magic_number str],
+      which are likely to come from the file being completely
+      different, and [Corrupted str], raised by headers that are the
+      (possibly empty) prefix of a valid magic number.
+
+      Unexpected errors correspond to valid magic numbers that are not
+      the one expected, either because it corresponds to a different
+      kind, or to a newer or older version. The helper function
+      [explain] will generate a textual explanation of the mismatch
+      between the expectation and the actual header, for use in error
+      messages.
+
+      @since 4.05.0
+  *)
+
+  (* This API could be improved in the following ways:
+
+     - It is conceivable that future compiler versions might support
+       several versions of a given object file format at the same time
+       (for example to provide some transition period). Then the
+       current_* API would be insufficient, as a range or set of
+       versions would be supported. *)
+
+  type version = int
+  type kind =
+      | Exec | Cmi | Cmo | Cma | Cmx | Cmxa | Cmxs | Cmt
+      | Ast_impl | Ast_intf
+
+  type raw = string
+  (** the type of raw magic numbers, such as "Caml1999O011" *)
+
+  type info = {
+    kind: kind;
+    version: version;
+  }
+
+  type parse_error =
+    | Corrupted of string
+    | Not_a_magic_number of string
+
+  exception Parse_error of parse_error
+
+  val parse : raw -> (info, parse_error) result
+  (** Parses a raw magic number *)
+
+  val read_info : in_channel -> (info, parse_error) result
+  (** Read a raw magic number from an input channel.
+
+      If the data read [str] is not a valid magic number, it can be
+      recovered from the [Truncated str | Not_a_magic_number str]
```

Comment:
`Truncated` should be `Corrupted`.

Reply:
In the end I went back to `Truncated`. Corrupted is used for files that do have a valid magic number but then fail when parsing the file along the expected format.